FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -5,38 +5,156 @@
 import CoreData
 
 public final class CoreDataFeedStore: FeedStore {
-	private static let modelName = "FeedStore"
+	private static let modelName = "CoreDataFeedStore"
+
 	private static let model = NSManagedObjectModel(name: modelName, in: Bundle(for: CoreDataFeedStore.self))
-	
+
 	private let container: NSPersistentContainer
 	private let context: NSManagedObjectContext
-	
+
 	struct ModelNotFound: Error {
 		let modelName: String
 	}
-	
+
 	public init(storeURL: URL) throws {
 		guard let model = CoreDataFeedStore.model else {
 			throw ModelNotFound(modelName: CoreDataFeedStore.modelName)
 		}
-		
+
 		container = try NSPersistentContainer.load(
 			name: CoreDataFeedStore.modelName,
 			model: model,
 			url: storeURL
 		)
 		context = container.newBackgroundContext()
 	}
-	
+
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		do {
+			if
+				let cache = try CoreDataFeedCache.fetch(context: context),
+				let feedItems = cache.feedItems,
+				let timestamp = cache.timestamp {
+				let feed = mapToLocalFeed(feedItems: feedItems)
+				completion(.found(feed: feed, timestamp: timestamp))
+			} else {
+				completion(.empty)
+			}
+		} catch {
+			completion(.failure(error))
+		}
+	}
+
+	private func deleteCache() throws {
+		try container.persistentStoreCoordinator.execute(CoreDataFeedCache.deleteRequest(), with: context)
+		try context.save()
 	}
-	
+
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		do {
+			try deleteCache()
+			try CoreDataFeedCache.insert(feed: feed, timestamp: timestamp, context: self.context)
+			try self.context.save()
+			completion(nil)
+		} catch {
+			completion(error)
```

Comment:
If there's an insertion error, the changes in the managed model aren't automatically reverted. They stay in memory.

To make the test pass, you need to roll back the changes.

```suggestion
			self.context.rollback()
			completion(error)
```